syntax = "proto3";

package jokinojo;

// Dosya paylaşımı için mesaj yapısı
message FileShare {
  enum DataType {
    NULL = 0;
    FILE_INFO = 1;    // Dosya hakkında bilgi (adı, boyutu, hash)
    PIECE = 2;        // Bir dosya parçası
    MISSING = 3;      // Eksik parçaların bilgisi
    MISSING_INFO = 4; // Dosya bilgisi eksikse
    FINISHED = 5;     // Gönderim tamamlandı
  }

  DataType datatype = 1;          // Mesaj türü
  optional string fileName = 2;   // Dosyanın adı (FILE_INFO için)
  optional int64 fileSize = 3;    // Dosyanın boyutu (byte olarak)
  optional string hash = 4;       // Dosyanın bütünlük kontrolü için hash (ör: SHA-256)

  // Dosya parçaları için alanlar
  optional int32 pieceIndex = 5;  // Parça numarası (PIECE için)
  optional bytes pieceData = 6;   // Dosya parçası verisi (PIECE için)

  // Eksik parçalar için alanlar
  repeated int32 missingPieces = 7; // Eksik parça numaralarının listesi (MISSING için)
}

message RequestData {
  enum DataType {
    NULL = 0;
    CREATE_ROOM = 1;
    JOIN_ROOM = 2;
    QUIT = 3;
    SYNC = 4;
    VIDEO_NAME = 5;
    READY = 6;
    CHAT = 7;
    FILE_SHARE = 8; // Yeni bir tür: Dosya paylaşımı
  }
  DataType dataType = 1;
  optional string username = 2;
  optional int32 roomId = 3;
  optional int32 timePosition = 4;
  optional bool resumed = 5;
  optional string videoName = 6;
  optional string chatMessage = 7;
  optional FileShare fileShare = 8; // FileShare mesajı eklendi
}

message ResponseData {
  enum DataType {
    NULL = 0;
    CREATE_ROOM = 1;
    JOIN_ROOM = 2;
    USER_LEFT = 3;
    SYNC = 4;
    VIDEO_NAME = 5;
    READY = 6;
    CHAT = 7;
    ERROR = 8;
    FILE_SHARE = 9; // Yeni bir tür: Dosya paylaşımı
  }
  DataType dataType = 1;
  optional int32 roomId = 2;
  repeated string usernames = 3;
  optional string username = 4;
  optional int32 timePosition = 5;
  optional bool resumed = 6;
  optional string videoName = 7;
  optional string chatMessage = 8;
  optional string errorMessage = 9;
  optional FileShare fileShare = 10; // FileShare mesajı eklendi
}
